server/image_pull.go

```diff
@@ -27,15 +27,57 @@ func (s *Server) PullImage(ctx context.Context, req *pb.PullImageRequest) (resp
 		image = img.Image
 	}
 
+	// We use the server's pullCache to record which images are currently being
+	// pulled. This allows for avoiding pulling the same image in parallel.
+	// Hence, if a given image is currently being pulled, we queue into the
+	// pullOperation's waitgroup and wait for the pulling goroutine to unblock
+	// us and re-use its results.
+	pullOp := &pullOperation{}
```

Comment:
Non-blocking: This could be `var pullOp *pullOperation`, initializing to `nil`; the heap allocation can be moved inside the `!pullInProcess` case.

Reply:
Nice idea!